lib/pdf_attachment_reporter.rb

```diff
@@ -0,0 +1,179 @@
+
+            if last_published_edition_with_attachment
+              pdf_owning_organisation = guess_organisation_owner_of_edition(last_published_edition_with_attachment)
+              first_published_edition_with_attachment = find_first_published_version_with_attachment(attachable_object.consultation, attachment.attachment_data)
+            end
+          end
+        else
+          last_published_edition_with_attachment = find_last_published_edition_with_attachment(attachable_object, attachment.attachment_data)
+
+          if last_published_edition_with_attachment
+            pdf_owning_organisation = guess_organisation_owner_of_edition(last_published_edition_with_attachment)
+            first_published_edition_with_attachment = find_first_published_version_with_attachment(attachable_object, attachment.attachment_data)
+          end
+        end
+
+        if pdf_owning_organisation && pdf_owning_organisation.live? &&
+            first_published_edition_with_attachment && first_published_edition_with_attachment.public_timestamp
+          PDFAttachmentData.new(pdf_owning_organisation.name, first_published_edition_with_attachment.public_timestamp)
+        end
+      end
+    end
+  end
+
+  def find_attachable_object(attachment)
+    attachable_type = attachment.attachable_type.constantize
+    attachable_type.find_by_id(attachment.attachable_id)
+  end
+
+  def find_last_published_edition_with_attachment(edition, attachment_data)
+    edition.document.editions.order('created_at DESC').detect do |ed|
+      Edition::POST_PUBLICATION_STATES.include?(ed.state) && ed.attachments.any? { |a| a.attachment_data_id == attachment_data.id }
+    end
+  end
+
+  def find_first_published_version_with_attachment(edition, attachment_data)
+    edition.document.editions.order('created_at ASC').detect do |ed|
+      Edition::POST_PUBLICATION_STATES.include?(ed.state) && ed.attachments.any? { |a| a.attachment_data_id == attachment_data.id }
+    end
+  end
+
+  def guess_organisation_owner_of_edition(edition)
+    if edition.organisations.first.class == WorldwideOrganisation
+      # Editions associated with a WWO are associated with one only
+      edition.organisations.first.sponsoring_organisation
+    else
+      organisation_ids = edition.organisations.map(&:id)
+
+      if any_organisations_unrelated?(edition, organisation_ids)
+        # If there is at least one unrelated organisation, assume the first organisation in the list is the owner
+        if edition.respond_to?(:lead_organisations)
+          edition.lead_organisations.find { |org| org.class == Organisation }
+        else
+          edition.organisations.find { |org| org.class == Organisation }
+        end
+      else
+        # If all organisations are related, use the parent organisation as the owner
+        edition.organisations.find do |org|
+          (org.parent_organisation_ids & organisation_ids).none?
+        end
```

Comment:
Unless rubocop says otherwise I think it can be useful to prefer `detect` over `find` when working with things that might be rails associations because it makes it clear you are using the ruby Enumerable API and not the rails active record API.  You've got 3 places here where you use `find` that you could change if you agree (and you use `detect` elsewhere so it'd be consistent).